atintegrators/VariableThinMPolePass.c

```diff
@@ -0,0 +1,274 @@
+/* VariableThinMPolePass
+   Accelerator Toolbox 
+   S.White simon.white@esrf.fr
+*/
+
+#include "atelem.c"
+#include "atlalib.c"
+#include "driftkick.c"
+
+#define TWOPI  6.28318530717959
+
+
+struct elemab
+{
+    double *Amplitude;
+    double Frequency;
+    double Phase;
+    int NSamples;
+    double *Func;
+};
+
+struct elem 
+{
+    double *PolynomA;
+    double *PolynomB;
+    struct elemab *ElemA;
+    struct elemab *ElemB;
+    int Seed;
+    int Mode;
+    int MaxOrder;
+    double *Ramps;
+};
+
+
+double randn(int seed)
+{
+  double u, v, w, mult;
+  static int initseed=1;
+
+  if(initseed){
+      srand(seed);
+      initseed=0;
+  }
+
+  do{
+      u = -1 + ((double) rand () / RAND_MAX) * 2;
+      v = -1 + ((double) rand () / RAND_MAX) * 2;
+      w = u*u + v*v;
+  }
+  while (w >= 1 || w == 0);
+  mult=sqrt ((-2 * log (w)) / w);
+  return u*mult;
+}
+
+double get_amp(double amp, double *ramps, double t)
+{
+    double ampt=amp;
+    if(ramps){
+        if(t<=ramps[0]){
+            ampt=0.0;
+        }else if(t<=ramps[1]){
+            ampt=amp*(t-ramps[0])/(ramps[1]-ramps[0]);
+        }else if(t<=ramps[2]){
+            ampt=amp;
+        }else if(t<=ramps[3]){
+            ampt=amp-amp*(t-ramps[2])/(ramps[3]-ramps[2]);
+        }else{
+            ampt=0.0;
+        }
+    }
+    return ampt;
+}
+
+double get_pol(struct elemab *elem, double *ramps, int mode,
+               double t, int turn, int seed, int order)
+{
+    int idx;
+    double ampt, freq, ph, val;
+    double *func;
+    double *amp = elem->Amplitude;
+    if(!amp){
+        return 0.0;
+    }
+    ampt=get_amp(amp[order],ramps,turn);
+    switch(mode){
+    case 0:
+        freq = elem->Frequency;
+        ph = elem->Phase;
+        ampt *= sin(TWOPI*freq*t+ph);
+        return ampt;
+    case 1:
+        val = randn(seed);
+        ampt *= val;
+        return ampt;
+    case 2:
+        func = elem->Func;
+        idx = turn%elem->NSamples;
+        ampt *= func[idx];
+        return ampt;
+    }
+}
+
+
+void VariableThinMPolePass(double *r, struct elem *Elem, double t0, int turn, int num_particles)
+{
+
+    int i, c;
+    double *r6;
+    double t = t0*turn;
+
+    int maxorder = Elem->MaxOrder;
+    double *pola = Elem->PolynomA;
+    double *polb = Elem->PolynomB;
+    int seed = Elem->Seed;
+    int mode = Elem->Mode;
+    struct elemab *ElemA = Elem->ElemA;
+    struct elemab *ElemB = Elem->ElemB;
+    double *ramps = Elem->Ramps;
+
+    for(i=0;i<maxorder+1;i++){
+        pola[i]=get_pol(ElemA, ramps, mode, t, turn, seed, i);
+        polb[i]=get_pol(ElemB, ramps, mode, t, turn, seed, i);
+    };
+

```

Comment:
This loop rewrites completely the `PolynomA/B` fields of the element. It is not recommended for a passmethod to modify its input. Can't we allocate locally `pola` and `polb`(`atMalloc`/`atFree`) rather than using the element data fields? Then `PolynomA` and `PolynomB` can be removed the element fields.

Reply:
I did implement it as you suggest to start but them re-introduce PolynomA/B because it was useful to keep track of their values for debugging.
I thought that users may find it useful as well to have access to these values so I left them in. They can be remove of course.
Is there a reason why a passmethod cannot change its element fields? We have several such passmethods for collective effects.

Reply:
OK, no problem. The only danger, and that's why the element attributes should better be "read-only" in the integrator, is that if a user modifies (shortens,  removes…) these attributes, the integrator may crash… Since `PolynomA` and `PolynomB` are not documented, there's a small risk. We can live with that !

Reply:
How do you handle this in standard elements. if `MaxOrder>length(PolynomA/B)` is crashes too right?